framework/src/main/java/org/checkerframework/framework/ajava/TypeAnnotationMover.java

```diff
@@ -0,0 +1,226 @@
+     * Returns the TypeElement for an annotation, or null if it cannot be found.
+     *
+     * <p>If {@code annotation} was listed in the file's imports, returns its value in {@link
+     * #allAnnotations}. If it wasn't imported but its element could still be found, adds the new
+     * TypeElement to {@link #allAnnotations} and returns it.
+     *
+     * @param annotation a JavaParser annotation
+     * @return the TypeElement for {@code annotation}, or null if it cannot be found
+     */
+    private @Nullable TypeElement getAnnotationDeclaration(AnnotationExpr annotation) {
+        @SuppressWarnings("signature") // https://tinyurl.com/cfissue/3094
+        @FullyQualifiedName String annoNameFq = annotation.getNameAsString();
+        TypeElement annoTypeElt = allAnnotations.get(annoNameFq);
+        if (annoTypeElt == null) {
+            annoTypeElt = elements.getTypeElement(annoNameFq);
+            if (annoTypeElt == null) {
+                // Not a supported annotation.
+                return null;
+            }
+            AnnotationFileParser.putAllNew(
+                    allAnnotations,
+                    AnnotationFileParser.createNameToAnnotationMap(
+                            Collections.singletonList(annoTypeElt)));
+        }
+
+        return annoTypeElt;
+    }
+
+    /**
+     * Returns if {@code annotation} could be a declaration annotation for {@code declarationType}.
+     * This would be the case if the annotation isn't recognized at all, or if it was recognized and
+     * has {@code declarationType} as one of its targets.
+     *
+     * @param annotation a JavaParser annotation expression
+     * @param declarationType the declaration type to check if {@code annotation} might be a
+     *     declaration annotation for
+     * @return false unless {@code annotation} definitely cannot be a declaration annotation for
+     *     {@code declarationType}
+     */
+    private boolean isPossiblyDeclarationAnnotation(
+            AnnotationExpr annotation, ElementType declarationType) {
+        TypeElement annotationType = getAnnotationDeclaration(annotation);
+        if (annotationType == null) {
+            return true;
+        }
+
+        return isPossiblyDeclarationAnnotation(annotationType, declarationType);
+    }
+
+    /**
+     * Returns whether the annotation represented by {@code annotationDeclaration} might be a
+     * declaration annotation for {@code declarationType}. This holds if {@code declarationType} is
+     * a target of the annotation, or if {@code ElementType.TYPE_USE} is not a target of the
+     * annotation.
+     *
+     * @param annotationDeclaration declaration for an annotation
+     * @param declarationType the declaration type to check if the annotation might be a declaration
+     *     annotation for
+     * @return true if {@code annotationDeclaration} contains {@code declarationType} as a target or
+     *     doesn't contain {@code ElementType.TYPE_USE} as a target
```

Comment:
I don't see why the second clause (after "or") is necessary.  Maybe the code should issue an error if the targets include neither `declarationType` nor `TYPE_USE`, though, as a sanity check.